### ISSUE TITLE
Logstash, datastream overwriting namespace

### ISSUE THREAD

**Issue by simlecp** [elastic/logstash#13792]
Hello everyone,

I am currently in a project to implement an ELK cluster under K8S.
The main objective is to centralize the logs of all our customers, while ensuring a partitioning of the logs according to the customer environments.

We are currently in the process of gaining expertise with Elastic solutions. In terms of our infrastructure, clients use Elastics agents, as well as Metricbeat and Filebeat beats.
The agents and beats are already deployed in the environments. As for the beats, they use a default configuration.

Our basic idea was to manage the partitioning of client environments through indexing in Logstash.
After several configuration tests, we decided to focus on indexing with datastreams.
Our goal was to partition client environments using data_stream_namespace (used in the creation of the datastream index) and spaces in kibana.
In our case, the problem comes from the fact that the beats use a default configuration (so the namespace is "Default").

So we have some questions :

1) Is it possible to overwrite the information in the data_stream fields in logstash to allow the namespace to be changed from "Default" to "Foo"? (We have done several tests but if it's possible to overwrite, then we have syntax problems)

2) Should all data_stream_* parameters be present in the Logstash output configuration ? We have tried to implement the data_stream_namespace. 
For logs/metrics that do not have a data_stream parameter, these have indeed retrieved the namespace configured in the Logstash output. As for the other data with data_stream fields. These have remained in "default".

3) What are the best practices for a multi-client environment ? The use of "classic" indexes or datastreams ?
What place does the fleet take in this type of environment ?

I thank you in advance for your help and wish you a good day !

Sincerely,
SD.

**Comment by simlecp** [comment]
You have here what we wanted to do with our output
`output {

  if  "foo" in [tags] {
    elasticsearch {
      hosts => ["contoso.com:443"]
      user => "user"
      password => "password"
      ssl => true
      data_stream => true
      data_stream_type => "%{[data_stream][type]}"
      data_stream_dataset => "%{[data_stream][dataset]}"
      data_stream_namespace => "foo"
    }
  }

  stdout { codec => rubydebug } 
}`

**Comment by sirReeall** [comment]
I was trying to do something similar, basically I wanted to use a field that is already defined as`data_stream_dataset`. 

I don't think sprintf is supported in the elasticsearch plugin.

I posted this up on the forum here

My example is as follows:

```
output {
  elasticsearch {
      hosts => "es01"
      data_stream => "true"
      data_stream_namespace => "%{[ticket_number]}"
  }
}
```

**Comment by simlecp** [comment]
Thank's for your answer ! 

We solved our problem ! :)

First of all, it's important to know that our approach does not take on board Fleet. Our infrastructure will switch to Fleet as soon as there is a release to use Fleet + Logstash.

Finally, we use Elastic agent integrations only. If we have understood correctly, in this use case, once the agent is configured, it manages the beats configuration itself. 
In the configuration of the Elastic agent, the datastream parameter is activated but we have left it by default and are doing updates/replace in Logstash filters, in order to modify the namespace and the dataset

For logs management according to namespaces, we have added a "customer" field to the Elasticsearch input: 

add_field => { "customer" => "example" }

Then we used a mutate filter to replace the namespace set to "default" with the name of our client. 
This allowed us to define roles and a partitioned view for each of our clients in Kibana

mutate {
  update => { "[data_stream][namespace]" => "%{[customer]}" }
}


kind regards,
SD
